### ISSUE TITLE
supervised learning

### ISSUE THREAD

**Issue by LOOKatKimhyein** [LOOKatKimhyein/study_ML#2]
### supervised learning
- labeled data
- direct feedback
- predict outcome or future
-> "classification"  "regression"

1) 일반화 선형 모델(GLM)
2) 의사결정트리(DT)
3) RANDOM FOREST
4) GBM/LGBM
5) DNN

**Comment by LOOKatKimhyein** [comment]
**1) 일반화 선형 모델(Generalized Linear Models)**
: 선형 회귀(linear regresison)의 발전된 형태
: 종속변수가 정규분포라는 가정을 할 수 없는 경우를 포함하는 선형 모형의 확장
ex1) 종속 변수가 이항, 다항 변수
ex2) 종속 변수가 제한된 count값으로 평균, 분산이 밀접하게 관련된 경우(하루 물 섭취량)
: 로지스틱, 포아송 등 함수를 이용해 연속형 변수로 변형

**Comment by LOOKatKimhyein** [comment]
**3) Random Forest**
- classificaton, regression 모두 가능(범주, 연속)
- overfitting이 적다(depth 가 너무 깊으면 과대적합가능성 有)
- missing value 처리용이
- black box로 결과를 설명할 수 없다
- bagging으로 low bias는 유지하고 high variance는 줄이는 방법
(bias - variance trade off)
- feature bagging(보통 sqrt(n))
- ***the last advantage is that the Random Forest classifier can be modeled for categorical values.
- bagging을 통해 각 트리들은 low correlation -> generalization 일반화 성능 향상
- 의사결정트리의 과적합, 성능 폭이 큰 문제를 보완
- *** 특징을 선택하는 이유는 배깅을 통해 얻은 트리들 간의 상관성에 있다. 즉, 한 개의 특징 또는 극소수의 특징들이 결과에 대해 강한 예측 성능을 지닌다면, 훈련 과정 중 여러 트리 노드에서 이러한 특징들이 중복되어 선택되고 결과적으로 트리들이 상관화되기 때문이다.
-> '신용등급'이 보험사기여부에 큰 영향을 미치는 변수라면, 여러 트리의 top split 에서 '신용등급'이 나와 트리 간 상관성이 높아지므로, feature bagging 을 통해 다양한 후보 컬럼에게 기회 제공하여 variance를 낮춘다


![image](https://user-images.githubusercontent.com/48271143/83670062-87b82780-a60d-11ea-8b22-5942ccc8c677.png)

1) 전체 m개 feature 중, 랜덤하게 k개를 추출 (k<<m)
2) 추출한 k개의 feature중, 최적의 split point인 노드를 계산
3) 계산된 노드를 시작으로 best split을 위한 하위 노드로 분리
4) 1-3 과정을 반복하여 terminal node까지 도달
5) 1-4과정을 반복하여 n개의 트리 생성


[https://towardsdatascience.com/understanding-random-forest-58381e0602d2](url)

[https://medium.com/@Synced/how-random-forest-algorithm-works-in-machine-learning-3c0fe15b6674](url)

**Comment by LOOKatKimhyein** [comment]
_"프로젝트에서 **NA포함된 학습데이터에 대한 RF의 처리 과정** 이슈 발생한 경험이 있다. 이에 대한 쉬운 설명을 공유하고자 한다"_
 [https://www.youtube.com/watch?v=sQ870aTKqiM   ](url)

전체적인 루틴은 '처음 guess를 점진적으로 발전시켜 적절한 guess값을 찾는다"

1.  RF에서 고려하는 missing value의 두 가지 유형

-  모델 생성에 쓰이는 original data set의 feature
-  test data set내의 NA

![image](https://user-images.githubusercontent.com/48271143/83791558-e5fc0d80-a6d4-11ea-874e-751bd99a8eef.png)

2. categorical 변수는 just 최빈값 & numeric 변수는 중앙값
na를 가진 sample의 lable이 'NO'이므로, 'NO'label인 샘플 중에서 계산한다.
![image](https://user-images.githubusercontent.com/48271143/83795123-f6fb4d80-a6d9-11ea-9da9-2cdc01383444.png)

3. 모든 데이터를 rf 전체 tree에 태운다
![image](https://user-images.githubusercontent.com/48271143/83792163-9ec24c80-a6d5-11ea-8846-14dbd13cb1d5.png)

4. 각 트리에서 leaf node가 같은 sample들을 1로 채우며 proximity matrix작성
![image](https://user-images.githubusercontent.com/48271143/83792857-a0d8db00-a6d6-11ea-8f82-fdb7836b245d.png)
![image](https://user-images.githubusercontent.com/48271143/83792904-b2ba7e00-a6d6-11ea-9d98-c9009d6613fc.png)

전체 tree 갯수로 나누어 확률 proximity 행렬을 만들고
이를 이용해 missing data를 고쳐나간다
![image](https://user-images.githubusercontent.com/48271143/83793017-e09fc280-a6d6-11ea-83b5-fc926ce017dc.png)

5. categorical 변수는 weighted frequency, numeric 변수는 weighted average(가중평균)을 이용한다

![image](https://user-images.githubusercontent.com/48271143/83793161-1f357d00-a6d7-11ea-9d6e-69507d7cb7ae.png)
![image](https://user-images.githubusercontent.com/48271143/83793239-3bd1b500-a6d7-11ea-9805-54b65c3ea56d.png)


![image](https://user-images.githubusercontent.com/48271143/83793317-59068380-a6d7-11ea-81cc-10fcf64b6507.png)


6. 'no', 198.5로 추측한 missing value를 교정한다.
7. missing value가 수렴할 때 까지 반복한다
[ 정리] =================================
1) RF를 만든다
2) DATA를 모델에 태운다
3) Proximity marix를 계산하여 missing value 재계산
=========================================

**Comment by LOOKatKimhyein** [comment]
### GBM

**Comment by LOOKatKimhyein** [comment]
**XGBoost**
- GBM은 residual을 줄이는 방향의 알고리즘으로 overfitting이 되기 쉽다
-> 이를 보완한 알고리즘이 XGBoost
- XGBoost = GBM + Regulation
![image](https://user-images.githubusercontent.com/48271143/83907523-8408da00-a7a0-11ea-96e2-c69ceb77905e.png)
![image](https://user-images.githubusercontent.com/48271143/83907554-92ef8c80-a7a0-11ea-9e3f-620d5cdb7597.png)

T는 트리의 최종 node의 갯수로, 커질수록 XGBoost loss function을 크게해 loss에 패널티를 주는 방식이다. 이로써 overfitting을 방지한다.

**Comment by LOOKatKimhyein** [comment]
### LightGBM
